pkg/nodeconfig/nodeconfig.go

```diff
@@ -288,6 +303,110 @@ func (nc *nodeConfig) Configure() error {
+		ignition.KubeletCACertPath: {},
+	}
+
+	ign, err := ignition.New(nc.client)
+	if err != nil {
+		return err
+	}
+	kubeletArgs, err := ign.GetKubeletArgs()
+	if err != nil {
+		return err
+	}
+	if _, ok := kubeletArgs[ignition.CloudConfigOption]; ok {
+		filesToTransfer[ignition.CloudConfigPath] = struct{}{}
+	}
+	// For each new file in the ignition file check if is a file we are interested in and, if so, decode its contents
+	for _, ignFile := range ign.GetFiles() {
+		if _, ok := filesToTransfer[ignFile.Node.Path]; ok {
+			if ignFile.Contents.Source == nil {
+				return errors.Errorf("could not process %s: File is empty", ignFile.Node.Path)
+			}
+			contents, err := url.QueryUnescape(*ignFile.Contents.Source)
+			if err != nil {
+				return errors.Wrapf(err, "could not decode %s", ignFile.Node.Path)
+			}
+			// Write to file on the Windows instance
+			fileName := filepath.Base(ignFile.Node.Path)
+			if err := nc.Windows.EnsureFileContent([]byte(contents), fileName, windows.K8sDir); err != nil {
+				return err
+			}
+			nc.log.V(1).Info("created", "file", windows.K8sDir+fileName)
+		}
+	}
+	return nil
+}
+
+// createBootstrapKubeconfig creates a kubeconfig for kubelet to initially communicate with the API server
+func (nc *nodeConfig) createBootstrapKubeconfig() error {
+	nodeBootstrapToken, err := nc.k8sclientset.CoreV1().Secrets(mcoNamespace).Get(context.TODO(), mcoBootstrapSecret,
+		meta.GetOptions{})
+	if err != nil {
+		return err
+	}
+	kubeconfig, err := generateKubeconfig(nodeBootstrapToken, nodeConfigCache.apiServerEndpoint)
+	if err != nil {
+		return err
+	}
+	kubeconfigData, err := json.Marshal(kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	dir, fileName := splitWindowsPath(windows.BootstrapKubeconfig)
+	if err := nc.Windows.EnsureFileContent(kubeconfigData, fileName, dir); err != nil {
+		return err
+	}
+	nc.log.V(1).Info("created", "file", windows.BootstrapKubeconfig)
+	return nil
+}
+
+// createKubeletConf creates config file for kubelet, with Windows specific configuration.
+func (nc *nodeConfig) createKubeletConf(clusterDNS string) error {
+	kubeletConfig := generateKubeletConfiguration(clusterDNS)
+	kubeletConfigData, err := json.Marshal(kubeletConfig)
+	if err != nil {
+		return err
+	}
+	// Appending this option is needed here instead of in the kubelet configuration object. Otherwise, when marshalling,
+	// the empty value will be omitted, so it would end up being incorrectly populated at service start time.
+	enforceNodeAllocatable := "\"enforceNodeAllocatable\":[]}"
+	// Replace last character ('}') with comma
+	kubeletConfigData[len(kubeletConfigData)-1] = ','
+	kubeletConfigData = []byte(string(kubeletConfigData) + enforceNodeAllocatable)
+
+	dir, fileName := splitWindowsPath(windows.KubeletConfigPath)
+	if err := nc.Windows.EnsureFileContent(kubeletConfigData, fileName, dir); err != nil {
+		return err
+	}
+	nc.log.V(1).Info("created", "file", windows.KubeletConfigPath)
+	return nil
+}
+
```

Comment:
object-oriented version
```suggestion
// createBootstrapKubeconfig creates a kubeconfig for kubelet to initially communicate with the API server
func (nc *nodeConfig) createBootstrapKubeconfig() error {
	nodeBootstrapToken, err := nc.k8sclientset.CoreV1().Secrets(mcoNamespace).Get(context.TODO(), mcoBootstrapSecret,
		meta.GetOptions{})
	if err != nil {
		return err
	}
	kubeconfig, err := generateKubeconfig(nodeBootstrapToken, nodeConfigCache.apiServerEndpoint)
	if err != nil {
		return err
	}
	kubeconfigData, err := json.Marshal(kubeconfig)
	if err != nil {
		return err
	}
	return nc.createFile(kubeconfigData, windows.BootstrapKubeconfig)
}

// createKubeletConf creates config file for kubelet, with Windows specific configuration.
func (nc *nodeConfig) createKubeletConf(clusterDNS string) error {
	kubeletConfig := generateKubeletConfiguration(clusterDNS)
	kubeletConfigData, err := json.Marshal(kubeletConfig)
	if err != nil {
		return err
	}
	// Appending this option is needed here instead of in the kubelet configuration object. Otherwise, when marshalling,
	// the empty value will be omitted, so it would end up being incorrectly populated at service start time.
	enforceNodeAllocatable := "\"enforceNodeAllocatable\":[]}"
	// Replace last character ('}') with comma
	kubeletConfigData[len(kubeletConfigData)-1] = ','
	kubeletConfigData = []byte(string(kubeletConfigData) + enforceNodeAllocatable)
	return nc.createFile(kubeletConfigData, windows.KubeletConfigPath)
}

func (nc *nodeConfig) createFile(jsonData []byte, path string) error {
	writer := JSONFileWriter{
		data:     jsonData,
		path:     windows.KubeletConfigPath,
		instance: nc.Windows,
	}
	if err := writer.write(); err != nil {
		return err
	}
	nc.log.V(1).Info("created", "file", windows.KubeletConfigPath)
	return nil
}

type JSONFileWriter struct {
	// data is the marshalled data to be used as file contents.
	data []byte
	// path is a Windows file path where the data will be written.
	path string
	// instance is the Windows instance to write the file to.
	instance windows.Windows
}

// write writes the data within the JSONFileWriter to the path on the given Windows instance. Creates the file if needed
func (w *JSONFileWriter) write() error {
	dir, fileName := splitWindowsPath(w.path)
	return w.instance.EnsureFileContent(w.data, fileName, dir)
}
```

Reply:
decided not to use the OO approach